File: frontend/src/lib/types/nistAiRmf.ts
export type NistControlStatus =
  | "met"
  | "partial"
  | "not_met"
  | "not_assessed"
  | "not_applicable";

export type NistFindingKind =
  | "satisfied"
  | "violation"
  | "alignment_gap"
  | "improvement"
  | "unchecked"
  | "out_of_scope";

export type NistComplianceStatus = "compliant" | "attention_needed" | "non_compliant";

export interface NistControlCatalogItem {
  id: string;
  function: string;
  category_id: string;
  category_title?: string;
  text: string;
  evidence_type: string;
  coverage: string;
  modules: string[];
  evaluator?: string | null;
  trustworthiness: string[];
  notes?: string | null;
}

export interface NistControlsCatalog {
  version: string;
  source?: string;
  source_url?: string;
  playbook_url?: string;
  profile: {
    id: string;
    name: string;
    description: string;
    type: string;
  };
  trustworthiness_characteristics: string[];
  control_count: number;
  controls: NistControlCatalogItem[];
}

export interface NistControlStatusItem {
  id: string;
  function: string;
  category_id?: string;
  text: string;
  evidence_type: string;
  coverage: string;
  modules: string[];
  trustworthiness: string[];
  status: NistControlStatus;
  finding_kind: NistFindingKind;
  evidence: string[];
  detail: Record<string, unknown>;
  notes?: string | null;
}

export interface NistFunctionSummary {
  met: number;
  partial: number;
  not_met: number;
  not_assessed: number;
  not_applicable: number;
  violations?: number;
  alignment_gaps?: number;
}

export interface NistCurrentProfile {
  profile_id?: string;
  profile_name?: string;
  framework_version: string;
  evaluated_at: string;
  alignment_score: number;
  compliance_status: NistComplianceStatus;
  summary: NistFunctionSummary & {
    total: number;
    automated_evaluations: number;
    violations: number;
    alignment_gaps: number;
    evaluated_total: number;
    progress_total: number;
    base_score: number;
    violation_deduction: number;
  };
  by_function: Record<string, NistFunctionSummary>;
  controls: NistControlStatusItem[];
  disclaimer: string;
}

export const NIST_FUNCTIONS = ["GOVERN", "MAP", "MEASURE", "MANAGE"] as const;

export const NIST_STATUS_LABELS: Record<NistControlStatus, string> = {
  met: "Satisfied",
  partial: "Partial",
  not_met: "Not met",
  not_assessed: "Not checked yet",
  not_applicable: "Out of scope",
};

export const NIST_FINDING_KIND_LABELS: Record<NistFindingKind, string> = {
  satisfied: "Satisfied",
  violation: "Violation",
  alignment_gap: "Setup gap",
  improvement: "Improve",
  unchecked: "Not checked",
  out_of_scope: "Out of scope",
};

export const NIST_COMPLIANCE_STATUS_LABELS: Record<NistComplianceStatus, string> = {
  compliant: "Compliant",
  attention_needed: "Setup in progress",
  non_compliant: "Action needed",
};

export const NIST_STATUS_COLORS: Record<NistControlStatus, string> = {
  met: "text-accent-green",
  partial: "text-accent-amber",
  not_met: "text-accent-red",
  not_assessed: "text-text-muted",
  not_applicable: "text-text-muted",
};

File: backend/app/schemas/nist_ai_rmf.py
"""NIST AI RMF API schemas."""

from typing import Any

from pydantic import BaseModel, Field


class NistControlCatalogItem(BaseModel):
    id: str
    function: str
    category_id: str
    category_title: str | None = None
    text: str
    evidence_type: str
    coverage: str
    modules: list[str] = Field(default_factory=list)
    evaluator: str | None = None
    trustworthiness: list[str] = Field(default_factory=list)
    notes: str | None = None


class NistControlsCatalogResponse(BaseModel):
    version: str
    source: str | None = None
    source_url: str | None = None
    playbook_url: str | None = None
    profile: dict[str, Any]
    trustworthiness_characteristics: list[str]
    control_count: int
    controls: list[NistControlCatalogItem]


class NistControlStatusItem(BaseModel):
    id: str
    function: str
    category_id: str | None = None
    text: str
    evidence_type: str
    coverage: str
    modules: list[str] = Field(default_factory=list)
    trustworthiness: list[str] = Field(default_factory=list)
    status: str
    evidence: list[str] = Field(default_factory=list)
    detail: dict[str, Any] = Field(default_factory=dict)
    notes: str | None = None


class NistFunctionSummary(BaseModel):
    met: int = 0
    partial: int = 0
    not_met: int = 0
    not_assessed: int = 0
    not_applicable: int = 0


class NistProfileSummary(BaseModel):
    total: int
    met: int
    partial: int
    not_met: int
    not_assessed: int
    not_applicable: int
    automated_evaluations: int


class NistCurrentProfileResponse(BaseModel):
    profile_id: str | None = None
    profile_name: str | None = None
    framework_version: str
    evaluated_at: str
    alignment_score: float
    summary: NistProfileSummary
    by_function: dict[str, NistFunctionSummary]
    controls: list[NistControlStatusItem]
    disclaimer: str
